### PR TITLE
feat: Add category create, get all, get single, and delete

### PR DESCRIPTION
**Pull request by minh-ht**

### CODE REVIEW

File: main/api/routes/category.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+
+@router.get("/", response_model=List[CategoryBatchResponseSchema], status_code=status.HTTP_200_OK)
+async def get_all_categories(session: Session = Depends(get_db_session)):
+    categories = category_crud.get_categories(session)
+    return categories
+
+
+@router.get("/{category_id}", response_model=CategoryResponseSchema, status_code=status.HTTP_200_OK)
+async def get_single_category(category_id, session: Session = Depends(get_db_session)):
```
Comment (ke-np) [inline main/api/routes/category.py]: Typing for `category_id`?

File: main/api/routes/category.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    category = category_crud.get_category_by_id(session, category_id)
+    if category is None:
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail="Cannot find the specified category"
+        )
```
Comment (ke-np) [inline main/api/routes/category.py]: Let's move this into a dependency so we can easily reuse it.

File: main/api/routes/category.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+async def create_category(
+    create_category_data: CategoryCreationRequestSchema,
+    session: Session = Depends(get_db_session),
+    user_id: Optional[int] = Depends(get_current_user_id)
+):
+    if user_id is None:
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="User needs to authenticate to create new category"
+        )
```
Comment (ke-np) [inline main/api/routes/category.py]: Move this into a dependency. We will need this check for other APIs soon.

File: main/api/routes/category.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Cannot delete the specified category"
+        )
+
+    if user_id != category.user_id:
+        raise HTTPException(
+            status_code=status.HTTP_403_FORBIDDEN,
+            detail="User does not have permission to delete this category"
+        )
```
Comment (ke-np) [inline main/api/routes/category.py]: This should be a dependency too.

File: main/schemas/category.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+class CategoryBase(BaseModel):
+    name: constr(min_length=1)
+
+    @validator("name")
+    def name_validator(cls, name_string):
+        msg = "Category name exceeds max length (50)"
+        if len(name_string) > 50:
+            raise ValueError(msg)
+        return name_string
```
Comment (ke-np) [inline main/schemas/category.py]: Why don't we do `name: constr(min_length=1, max_length=50)`?

File: main/schemas/category.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+    class Config:
+        orm_mode = True
```
Comment (ke-np) [inline main/schemas/category.py]: Let's have a base class for this as we're gonna reuse it for item schemas as well.

File: main/services/database/category_crud.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
+from typing import List
```
Comment (ke-np) [inline main/services/database/category_crud.py]: `main/services/category.py` is good enough. What convention are you following here?

File: main/services/database/category_crud.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+from sqlalchemy.orm import Session
+
+from main.models.category import Category
+from main.schemas.category import CategoryBatchResponseSchema
+
+
+def get_category_by_id(session: Session, category_id: str) -> Category:
+    category = session.query(Category).filter_by(id=category_id).one_or_none()
+    return category
```
Comment (ke-np) [inline main/services/database/category_crud.py]: There is a better API to get a record using its primary key. Please research & use that one instead.

File: main/services/database/category_crud.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
+from typing import List
+
+from sqlalchemy.orm import Session
+
+from main.models.category import Category
+from main.schemas.category import CategoryBatchResponseSchema
+
+
+def get_category_by_id(session: Session, category_id: str) -> Category:
```
Comment (ke-np) [inline main/services/database/category_crud.py]: Return typing should be `Optional[Category]`.

File: main/services/database/category_crud.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    category = session.query(Category).filter_by(id=category_id).one_or_none()
+    return category
+
+
+def get_category_by_name(session: Session, name: str) -> Category:
+    category = session.query(Category).filter_by(name=name).one_or_none()
+    return category
+
+
+def get_categories(session: Session) -> List[CategoryBatchResponseSchema]:
```
Comment (ke-np) [inline main/services/database/category_crud.py]: Why not `List[Category]`?

File: main/services/database/category_crud.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+
+def create_category(session: Session, name, description, user_id: int) -> None:
+    category = Category(
+        name=name,
+        description=description,
+        user_id=user_id
+    )
+    session.add(category)
+    session.commit()
```
Comment (ke-np) [inline main/services/database/category_crud.py]: Missing typing for name & description. This function is not well-designed, it should return the created category.

File: main/services/database/category_crud.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+
+def delete_category(session: Session, category: Category) -> None:
+    session.delete(category)
+    session.commit()
```
Comment (ke-np) [inline main/services/database/category_crud.py]: It's better to support deleting a record using its ID.

**Review by ke-np — CHANGES_REQUESTED** [review]
Also, please make use of async APIs for SQLAlchemy.

File: main/api/dependencies/auth.py [hunk 13]
```diff
@@ -20,7 +22,7 @@ async def require_authenticated_user(
 
     try:
         token = http_authorization_credentials.credentials
-        payload = jwt.decode(token, settings.JWT_SECRET_KEY, settings.ALGORITHM)
+        payload = jwt.decode(token, settings.JWT_SECRET_KEY, settings.JWT_ALGORITHM)
```
Comment (duongleh) [inline main/api/dependencies/auth.py]: Create a function in `auth` service for decoding token

File: main/api/dependencies/auth.py [hunk 14]
```diff
@@ -30,3 +32,12 @@ async def require_authenticated_user(
...
     return user
+
+
+async def require_permission_on_category(
+    user: UserModel = Depends(require_authenticated_user), category: CategoryModel = Depends(get_category_by_id)
+) -> None:
+    if category.user_id != user.id:
+        raise HTTPException(
+            status_code=status.HTTP_403_FORBIDDEN, detail="User does not have permission to perform this action"
+        )
```
Comment (duongleh) [inline main/api/dependencies/auth.py]: Custom exception

File: main/api/dependencies/category.py [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+from main.services import category as category_service
+
+
+async def get_category_by_id(category_id: int, session: AsyncSession = Depends(get_database_session)) -> CategoryModel:
+    category = await category_service.get_category_by_id(session, category_id)
+    if category is None:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Cannot find the specified category",
+        )
```
Comment (duongleh) [inline main/api/dependencies/category.py]: Should this be not found exception? Additionally, create a reusable not found exception.

File: main/api/routes/category.py [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    create_category_data: CategoryCreationRequestSchema,
+    session: AsyncSession = Depends(get_database_session),
+    user: UserModel = Depends(require_authenticated_user),
+):
+    category = await category_service.get_category_by_name(session, create_category_data.name)
+    if category:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Category name already exists",
+        )
```
Comment (duongleh) [inline main/api/routes/category.py]: Custom exception

**Review by duongleh — CHANGES_REQUESTED** [review]

File: main/services/category.py [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    category = CategoryModel(name=name, description=description, user_id=user_id)
+    session.add(category)
+    await session.commit()
+    return category
+
+
+async def delete_category(session: AsyncSession, category_id: int) -> None:
+    category = await session.get(CategoryModel, category_id)
+    if category is None:
+        raise NoEntityException
```
Comment (duong-le) [inline main/services/category.py:39]: `NoEntityException` or `NoEntityException()`?

**Review by duongleh — APPROVED** [review]

**Pull request merged**